src/crypto/backend/rust.rs

```diff
@@ -0,0 +1,110 @@
+
+use super::interface;
+
+pub(crate) mod ec;
+pub(crate) mod hmac;
+pub(crate) mod okp;
+pub(crate) mod rsa;
+
+// TODO: remove the `cfg_attr` once the RustCrypto crates implement
+// the core::error::Error trait.
+
+/// The errors that can be produced by the rust crypto backend.
+#[derive(Debug, Error)]
+pub(crate) enum BackendError {
+    /// The error returned if the key is invalid.
+    #[error("invalid key length")]
+    InvalidLength,
+
+    /// RSA operation failed.
+    #[cfg_attr(feature = "std", error("an RSA operation failed"))]
+    #[cfg_attr(not(feature = "std"), error("an RSA operation failed: {0}"))]
+    Rsa(#[cfg_attr(feature = "std", source)] ::rsa::errors::Error),
+
+    /// Error of the `elliptic_curve` crate.
+    #[cfg_attr(feature = "std", error("an elliptic curve operation failed"))]
+    #[cfg_attr(not(feature = "std"), error("an elliptic curve operation failed: {0}"))]
+    EllipticCurve(#[cfg_attr(feature = "std", source)] ::elliptic_curve::Error),
+
+    /// Error of the `ecdsa` crate.
+    #[cfg_attr(feature = "std", error("an ECDSA operation failed"))]
+    #[cfg_attr(not(feature = "std"), error("an ECDSA operation failed: {0}"))]
+    Ecdsa(#[cfg_attr(feature = "std", source)] ::ecdsa::Error),
+
+    /// Error of the `ed25519-dalek` crate.
+    #[cfg_attr(feature = "std", error("an ED25519 operation failed"))]
+    #[cfg_attr(not(feature = "std"), error("an ED25519 operation failed: {0}"))]
+    Ed25519(#[cfg_attr(feature = "std", source)] ed25519_dalek::SignatureError),
+
+    /// The amount of bytes for an EC point is invalid.
+    #[error("invalid EC point length, expected {expected}, got {actual}")]
+    InvalidEcPoint { expected: usize, actual: usize },
+
+    /// The coordinates did not form a valid key.
+    #[error("invalid EC key")]
+    InvalidEcKey,
+
+    /// The curve type is not supported by this backend.
+    #[error("curve '{0}' not supported by this backend")]
+    CurveNotSupported(&'static str),
+
+    #[error("RSA key expected to have exactly 2 prime numbers")]
+    RsaTwoPrimes,
+}
+
+impl From<digest::InvalidLength> for BackendError {
+    fn from(_: digest::InvalidLength) -> Self {
+        Self::InvalidLength
+    }
+}
+
+impl From<::rsa::errors::Error> for BackendError {
+    fn from(x: ::rsa::errors::Error) -> Self {
+        Self::Rsa(x)
+    }
+}
+
+/// The [RustCrypto] based backend.
+///
+/// [RustCrypto]: https://github.com/RustCrypto
+#[derive(Debug)]
+pub(crate) enum Backend {}
+
+impl interface::Backend for Backend {
+    type EcPrivateKey = ec::PrivateKey;
+    type EcPublicKey = ec::PublicKey;
+    type EdPrivateKey = okp::PrivateKey;
+    type EdPublicKey = okp::PublicKey;
+    type Error = BackendError;
+    type HmacKey = hmac::Key;
+    type RsaPrivateKey = rsa::PrivateKey;
+    type RsaPublicKey = rsa::PublicKey;
+
+    fn fill_random(buf: &mut [u8]) -> Result<(), Self::Error> {
+        use rand_core::OsRng;
+
+        OsRng.fill_bytes(buf);
+        Ok(())
```

Comment:
Does this work on no_std? Btw rand_core is outdatd, latest is 0.9, fill_bytes is replaced by [`try_fill_bytes`](https://docs.rs/rand_core/0.9.3/rand_core/struct.OsRng.html#method.try_fill_bytes)

Reply:
The problem is that all of the RustCrypto crates use `rand_core = "0.6.4"`, so we must use that too to have the correct types when generating keys.

Reply:
Also yes this works in `no_std` as long as one specifies a backend for the `getrandom` crate.

Reply:
Why don't we let the user provide the Rng like we did with the `generate` methods on the key types? e.g. https://docs.rs/jose/0.0.1/jose/jwk/ec/p256/struct.P256PrivateKey.html